app/controllers/concerns/unconfirmed_user_concern.rb

```diff
@@ -29,24 +29,10 @@ def track_user_already_confirmed_event
   end
 
   def stop_if_invalid_token
-    return if @email_address.present?
-    hash = {
-      success: false,
-      errors: { confirmation_token: [t('errors.messages.confirmation_invalid_token')] },
-      user_id: nil,
-    }
-    analytics.track_event(Analytics::USER_REGISTRATION_EMAIL_CONFIRMATION, hash)
-    process_unsuccessful_confirmation
-  end
-
-  def process_confirmation
     result = email_confirmation_token_validator.submit
```

Comment:
putting this validation check in the `stop_if_invalid_token` is the key part of the change to ensure that tokens are fully validated, rather than only checking if `@email_address` is present